// @linaria/logger/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.error = exports.warn = exports.info = exports.debug = void 0;

var _debug = _interopRequireDefault(require("debug"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const levels = ['error', 'warn', 'info', 'debug'];
const currentLevel = levels.indexOf(process.env.LINARIA_LOG || 'error');
const linariaLogger = (0, _debug.default)('linaria');
const loggers = new Map();

function gerOrCreate(namespace) {
  if (!namespace) return linariaLogger;
  const lastIndexOf = namespace.lastIndexOf(':');

  if (!loggers.has(namespace)) {
    loggers.set(namespace, gerOrCreate(namespace.substr(0, lastIndexOf)).extend(namespace.substr(lastIndexOf + 1)));
  }

  return loggers.get(namespace);
}

const format = text => {
  if (typeof text === 'string') {
    return text.replace(/\n/g, '\n\t');
  }

  return text;
};

function log(level, namespaces, arg1, ...restArgs) {
  if (currentLevel < levels.indexOf(level)) {
    return;
  }

  const logger = gerOrCreate(namespaces);
  if (!logger.enabled) return;

  if (typeof arg1 === 'function') {
    const text = arg1();

    if (text) {
      logger('', format(text), ...restArgs);
    }

    return;
  }

  logger('', format(arg1), ...restArgs);
}

const debug = log.bind(null, 'debug');
exports.debug = debug;
const info = log.bind(null, 'info');
exports.info = info;
const warn = log.bind(null, 'warn');
exports.warn = warn;
const error = log.bind(null, 'error');
exports.error = error;
